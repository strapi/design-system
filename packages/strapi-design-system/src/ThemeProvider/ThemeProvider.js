import * as React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider as StyledThemeProvider, createGlobalStyle } from 'styled-components';

const GlobalStyle = createGlobalStyle`
 /* http://meyerweb.com/eric/tools/css/reset/ 
   v2.0 | 20110126
   License: none (public domain)
*/
  html,
  body,
  div,
  span,
  applet,
  object,
  iframe,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6,
  p,
  blockquote,
  pre,
  a,
  abbr,
  acronym,
  address,
  big,
  cite,
  code,
  del,
  dfn,
  em,
  img,
  ins,
  kbd,
  q,
  s,
  samp,
  small,
  strike,
  strong,
  sub,
  sup,
  tt,
  var,
  b,
  u,
  i,
  center,
  dl,
  dt,
  dd,
  ol,
  ul,
  li,
  fieldset,
  form,
  label,
  legend,
  table,
  caption,
  tbody,
  tfoot,
  thead,
  tr,
  th,
  td,
  article,
  aside,
  canvas,
  details,
  embed,
  figure,
  figcaption,
  footer,
  header,
  hgroup,
  menu,
  nav,
  output,
  ruby,
  section,
  summary,
  time,
  mark,
  audio,
  video {
    margin: 0;
    padding: 0;
    border: 0;
    font-size: 100%;
    font: inherit;
    vertical-align: baseline;
  }
  /* HTML5 display-role reset for older browsers */
  article,
  aside,
  details,
  figcaption,
  figure,
  footer,
  header,
  hgroup,
  menu,
  nav,
  section {
    display: block;
  }
  body {
    line-height: 1;
  }
  ol,
  ul {
    list-style: none;
  }
  blockquote,
  q {
    quotes: none;
  }
  blockquote:before,
  blockquote:after,
  q:before,
  q:after {
    content: "";
    content: none;
  }
  table {
    border-collapse: collapse;
    border-spacing: 0;
  }
  /* My styles */
  *,
  *:before,
  *:after {
    box-sizing: border-box;
  }

  *:focus-visible {
    box-shadow: 0px 0px 6px ${(props) => props.theme.colors.primary500};
    // Windows High Contrast won't see the bow shadow but transparent will
    // be replaced by a color
    // See https://sarahmhigley.com/writing/whcm-quick-tips/ for more details
    outline: 4px solid transparent;
  }
  /* Focusing the button with a mouse, touch, or stylus will show a subtle drop shadow. */
  *:focus:not(:focus-visible) {
    outline: none;
  }

  body,html{
    height: 100%;
  }
  body {
    font-family: --apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
  }
`;

export const ThemeProvider = ({ children, theme }) => {
  return (
    <StyledThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
    </StyledThemeProvider>
  );
};

ThemeProvider.propTypes = {
  children: PropTypes.node.isRequired,
  theme: PropTypes.object.isRequired,
};
