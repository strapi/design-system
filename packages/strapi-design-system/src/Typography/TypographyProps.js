import React from 'react';
import PropTypes from 'prop-types';
import { TEXT_VARIANTS, OMEGA } from './constants';

export const TypographyProps = (props) => <div {...props} />;

export const typographyDefaultProps = {
  ellipsis: false,
  fontWeight: undefined,
  fontSize: undefined,
  lineHeight: undefined,
  textColor: undefined,
  textAlign: undefined,
  textTransform: undefined,
  variant: OMEGA,
};

export const typographyPropTypes = {
  ellipsis: PropTypes.bool,
  fontSize: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  fontWeight: PropTypes.string,
  lineHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  textAlign: PropTypes.string,
  textColor: PropTypes.string,
  textTransform: PropTypes.string,
  variant: PropTypes.oneOf(TEXT_VARIANTS),
};

TypographyProps.defaultProps = typographyDefaultProps;
TypographyProps.propTypes = typographyPropTypes;
