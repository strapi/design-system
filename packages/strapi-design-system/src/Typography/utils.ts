import { DefaultTheme } from 'styled-components';

import { ALPHA, BETA, DELTA, EPSILON, OMEGA, PI, SIGMA, TEXT_VARIANTS } from './constants';

export const ellipsisStyle = ({ ellipsis = false }: { ellipsis?: boolean }) =>
  ellipsis &&
  `
    display: block;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  `;

export const variantStyle = ({
  variant = OMEGA,
  theme,
}: {
  variant?: (typeof TEXT_VARIANTS)[number];
  theme: DefaultTheme;
}) => {
  switch (variant) {
    case ALPHA: {
      return `
        font-weight: ${theme.fontWeights.bold};
        font-size: ${theme.fontSizes[5]};
        line-height: ${theme.lineHeights[2]};
      `;
    }
    case BETA: {
      return `
        font-weight: ${theme.fontWeights.bold};
        font-size: ${theme.fontSizes[4]};
        line-height: ${theme.lineHeights[1]};
      `;
    }
    case DELTA: {
      return `
        font-weight: ${theme.fontWeights.semiBold};
        font-size: ${theme.fontSizes[3]};
        line-height: ${theme.lineHeights[2]};
      `;
    }
    case EPSILON: {
      return `
        font-size: ${theme.fontSizes[3]};
        line-height: ${theme.lineHeights[6]};
      `;
    }
    case OMEGA: {
      return `
        font-size: ${theme.fontSizes[2]};
        line-height: ${theme.lineHeights[4]};
      `;
    }
    case PI: {
      return `
        font-size: ${theme.fontSizes[1]};
        line-height: ${theme.lineHeights[3]};
      `;
    }
    case SIGMA: {
      return `
        font-weight: ${theme.fontWeights.bold};
        font-size: ${theme.fontSizes[0]};
        line-height: ${theme.lineHeights[5]};
        text-transform: uppercase;
      `;
    }
    default: {
      return `
        font-size: ${theme.fontSizes[2]};
      `;
    }
  }
};
