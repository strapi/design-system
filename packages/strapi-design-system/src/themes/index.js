export * from './lightTheme';
export * from './darkTheme';
export * from './extendTheme';
export * from './utils';
