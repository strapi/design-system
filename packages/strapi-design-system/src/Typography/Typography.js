import styled from 'styled-components';
import { ellipsisStyle, variantStyle, handleColor } from './utils';
import { typographyDefaultProps, typographyPropTypes } from './TypographyProps';

const transientProps = {
  fontSize: true,
  fontWeight: true,
};

export const Typography = styled.span.withConfig({
  shouldForwardProp: (prop, defPropValFN) => !transientProps[prop] && defPropValFN(prop),
})`
  ${variantStyle}
  ${ellipsisStyle}

  // These properties need to come after {variantStyle}, because they might
  // overwrite a variant attribute
  font-weight: ${({ theme, fontWeight }) => theme.fontWeights[fontWeight]};
  font-size: ${({ theme, fontSize }) => theme.fontSizes[fontSize]};
  line-height: ${({ theme, lineHeight }) => theme.lineHeights[lineHeight]};
  color: ${handleColor};
  text-align: ${({ textAlign }) => textAlign};
  text-transform: ${({ textTransform }) => textTransform};
`;

Typography.defaultProps = typographyDefaultProps;
Typography.propTypes = typographyPropTypes;
