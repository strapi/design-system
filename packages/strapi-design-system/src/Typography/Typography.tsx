import * as React from 'react';

import styled, { CSSProperties, DefaultTheme } from 'styled-components';

import { TEXT_VARIANTS } from './constants';
import { ellipsisStyle, variantStyle } from './utils';
import { extractStyleFromTheme } from '../helpers/theme';

const transientProps: Partial<Record<keyof TypographyProps, boolean>> = {
  fontSize: true,
  fontWeight: true,
};

type DefaultThemeOrCSSProp<T extends keyof DefaultTheme, K extends keyof CSSProperties> =
  | keyof DefaultTheme[T]
  | CSSProperties[K];

export interface TypographyProps<TElement extends HTMLElement = HTMLSpanElement>
  extends React.HTMLAttributes<TElement> {
  as?: string | React.ComponentType<any>;
  forwardedAs?: string | React.ComponentType<any>;
  children?: React.ReactNode;
  ellipsis?: boolean;
  fontSize?: keyof DefaultTheme['fontSizes'];
  fontWeight?: keyof DefaultTheme['fontWeights'];
  lineHeight?: DefaultThemeOrCSSProp<'lineHeights', 'lineHeight'>;
  textAlign?: CSSProperties['textAlign'];
  textColor?: keyof DefaultTheme['colors'];
  textDecoration?: CSSProperties['textDecoration'];
  textTransform?: CSSProperties['textTransform'];
  variant?: (typeof TEXT_VARIANTS)[number];
}

export const Typography = styled.span.withConfig<TypographyProps>({
  shouldForwardProp: (prop, defPropValFN) => !transientProps[prop as keyof TypographyProps] && defPropValFN(prop),
})`
  ${variantStyle}
  ${ellipsisStyle}

  // These properties need to come after {variantStyle}, because they might
  // overwrite a variant attribute
  font-weight: ${({ theme, fontWeight }) => extractStyleFromTheme(theme.fontWeights, fontWeight, undefined)};
  font-size: ${({ theme, fontSize }) => extractStyleFromTheme(theme.fontSizes, fontSize, undefined)};
  line-height: ${({ theme, lineHeight }) => extractStyleFromTheme(theme.lineHeights, lineHeight, lineHeight)};
  color: ${({ theme, textColor }) => theme.colors[textColor || 'neutral800']};
  text-align: ${({ textAlign }) => textAlign};
  text-decoration: ${({ textDecoration }) => textDecoration};
  text-transform: ${({ textTransform }) => textTransform};
`;
