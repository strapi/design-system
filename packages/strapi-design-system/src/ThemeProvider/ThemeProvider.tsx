/**
 * TODO: remove this in v2 in favour of DesignSystemProvider, but this is rather convienient for now.
 */
import * as React from 'react';

import { ThemeProvider as StyledThemeProvider, createGlobalStyle, DefaultTheme } from 'styled-components';

import { LiveRegions } from '../LiveRegions/LiveRegions';
import { lightTheme } from '../themes';

const GlobalStyle = createGlobalStyle`
 /* http://meyerweb.com/eric/tools/css/reset/ 
   v2.0 | 20110126
   License: none (public domain)
*/
  html,
  body,
  div,
  span,
  applet,
  object,
  iframe,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6,
  p,
  blockquote,
  pre,
  a,
  abbr,
  acronym,
  address,
  big,
  cite,
  code,
  del,
  dfn,
  em,
  img,
  ins,
  kbd,
  q,
  s,
  samp,
  small,
  strike,
  strong,
  sub,
  sup,
  tt,
  var,
  b,
  u,
  i,
  center,
  dl,
  dt,
  dd,
  ol,
  ul,
  li,
  fieldset,
  form,
  label,
  legend,
  table,
  caption,
  tbody,
  tfoot,
  thead,
  tr,
  th,
  td,
  article,
  aside,
  canvas,
  details,
  embed,
  figure,
  figcaption,
  footer,
  header,
  hgroup,
  menu,
  nav,
  output,
  ruby,
  section,
  summary,
  time,
  mark,
  audio,
  video {
    margin: 0;
    padding: 0;
    border: 0;
    font-size: 100%;
    font: inherit;
    vertical-align: baseline;
  }
  /* HTML5 display-role reset for older browsers */
  article,
  aside,
  details,
  figcaption,
  figure,
  footer,
  header,
  hgroup,
  menu,
  nav,
  section {
    display: block;
  }
  body {
    line-height: 1;
  }
  ol,
  ul {
    list-style: none;
  }
  blockquote,
  q {
    quotes: none;
  }
  blockquote:before,
  blockquote:after,
  q:before,
  q:after {
    content: "";
    content: none;
  }
  table {
    border-collapse: collapse;
    border-spacing: 0;
  }
  /* My styles */
  *,
  *:before,
  *:after {
    box-sizing: border-box;
  }

  *:focus-visible {
    outline: 2px solid ${({ theme }) => theme.colors.primary600};
    outline-offset: 2px;
  }

  /* Focusing the button with a mouse, touch, or stylus will show a subtle drop shadow. */
  *:focus:not(:focus-visible) {
    outline: none;
  }

  body,html{
    height: 100%;
  }
  body {
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
  }

  button {
    margin: 0;
    padding: 0;
    border: none;
    background: none;
    color: inherit;
    font-size: 100%;
    cursor: pointer;
    font: inherit;
  }

  textarea,input {
    font: inherit;
  }

  .lock-body-scroll {
    height: 100vh;
    overflow-y: hidden;
  }
`;

export interface ThemeProviderProps {
  children: React.ReactNode;
  theme?: DefaultTheme;
}

export const ThemeProvider = ({ children, theme = lightTheme }: ThemeProviderProps) => {
  return (
    <StyledThemeProvider theme={theme}>
      {children}
      <LiveRegions />
      <GlobalStyle />
    </StyledThemeProvider>
  );
};
