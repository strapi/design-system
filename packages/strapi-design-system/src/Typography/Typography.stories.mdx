<!--- Typography.stories.mdx --->

import { Meta, Story, Canvas } from '@storybook/addon-docs';
import { ArgsTable } from '@storybook/addon-docs';
import { Typography } from './Typography';
import { TypographyProps } from './TypographyProps';
import { Box } from '../Box';

<Meta title="Design System/Technical Components/Typography" component={Typography} />

# Text

This is the doc of the different Text / Typography components.

[View source](https://github.com/strapi/design-system/tree/main/packages/strapi-design-system/src/Typography)

## Imports

```js
import { Typography } from '@strapi/design-system';
```

## Usage of the different variants

The main way to display text content on the screen

<Canvas>
  <Story name="variants">
    <Box paddingTop={2} paddingLeft={4}>
      <Typography variant="alpha">Alpha - header 1</Typography>
    </Box>
    <Box paddingTop={2} paddingLeft={4}>
      <Typography variant="beta">Beta - header 2</Typography>
    </Box>
    <Box paddingTop={2} paddingLeft={4}>
      <Typography variant="delta">Delta - header 3</Typography>
    </Box>
    <Box paddingTop={2} paddingLeft={4}>
      <Typography variant="epsilon">Epsilon - subtitle</Typography>
    </Box>
    <Box paddingTop={2} paddingLeft={4}>
      <Typography variant="omega">Omega - body</Typography>
    </Box>
    <Box paddingTop={2} paddingLeft={4}>
      <Typography variant="omega" fontWeight="semiBold">
        Omega semibold - body highlight
      </Typography>
    </Box>
    <Box paddingTop={2} paddingLeft={4}>
      <Typography variant="omega" fontWeight="bold">
        Omega bold - button text
      </Typography>
    </Box>
    <Box paddingTop={2} paddingLeft={4}>
      <Typography variant="pi">Pi - small text</Typography>
    </Box>
    <Box paddingTop={2} paddingLeft={4}>
      <Typography variant="pi" fontWeight="bold">
        Pi bold - small button text
      </Typography>
    </Box>
    <Box paddingTop={2} paddingLeft={4}>
      <Typography variant="sigma">Sigma - table label</Typography>
    </Box>
    <Box paddingTop={2} paddingLeft={4} width="200px">
      <Typography ellipsis>Ellipsis example - long text - long text</Typography>
    </Box>
  </Story>
</Canvas>

### Ellipsis

By using the prop `ellipsis` it is possible to limit the text to one line and to signal to the user there is more to it.

<Canvas>
  <Story name="ellipsis">
    <Box maxWidth="150px" borderStyle="solid" borderWidth="1px" borderColor="neutral200">
      <Typography ellipsis>Loong text, that will be cut off</Typography>
    </Box>
  </Story>
</Canvas>

## Props

<ArgsTable of={TypographyProps} />
