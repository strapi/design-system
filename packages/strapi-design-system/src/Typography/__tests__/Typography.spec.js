import * as React from 'react';
import { render } from '@testing-library/react';
import { Typography } from '../Typography';
import { ThemeProvider } from '../../ThemeProvider';
import { lightTheme } from '../../themes';

const setup = (props) =>
  render(
    <ThemeProvider theme={lightTheme}>
      <Typography {...props} />
    </ThemeProvider>,
  );

describe('Typography', () => {
  test('textAlign', async () => {
    const { container } = setup({
      children: 'Test',
      textAlign: 'center',
    });

    expect(container).toMatchSnapshot();
  });

  test('textTransform', () => {
    const { container } = setup({
      children: 'Test',
      textTransform: 'uppercase',
    });

    expect(container).toMatchSnapshot();
  });
});
