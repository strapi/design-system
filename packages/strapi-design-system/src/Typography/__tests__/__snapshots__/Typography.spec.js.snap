// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography textAlign 1`] = `
.c0 {
  font-size: 0.875rem;
  line-height: 1.43;
  color: #32324d;
  text-align: center;
}

.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

<div>
  <span
    class="c0"
  >
    Test
  </span>
  <div
    class="c1"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`Typography textTransform 1`] = `
.c0 {
  font-size: 0.875rem;
  line-height: 1.43;
  color: #32324d;
  text-transform: uppercase;
}

.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

<div>
  <span
    class="c0"
  >
    Test
  </span>
  <div
    class="c1"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
